### DISCUSSION TITLE
(How) Is this doable with type inference?

### DISCUSSION THREAD

**Discussion by oliversaggau** [dotnet/roslyn#65809] [Q&A]
I have different methods in a project using .NET Framework (i.e. C# 7.3, though I also get the same behavior with .NET 7 and C# 11 which is not an option for my real project though, at least not yet)
```
public static StringAssertions AssertThat(string value)
{
    return new StringAssertions(value);
}

public static NumberAssertions<int> AssertThat(int value)
{
    return new NumberAssertions<int>(value);
}

[...]

public static EnumerableAssertions<T> AssertThat<T>(IEnumerable<T> collection)
{
    return new EnumerableAssertions<T>(collection);
}

public static ObjectAssertions<T> AssertThat<T>(T value) where T : class // generic fallback for "everything else" (that is a reference type)
{
    return new ObjectAssertions<T>(value);
}
```
But the following code resolves to the generic fallback instead of the overload with `IEnumerable<T>`
```
List<any-value-type-or-reference-type> list = null;
AssertThat(list);
```
I would haved guessed that `IEnumerable<T>` would be more specific than the "catch-all" `where T : class` for an argument `List<T>`. My first question would be, why isn't this the case? I tried to read C# specification paragraph "11.6.3 Type inference" but honestly couldn't find anything (but that's probably just me not having understood it correctly)

What does work though is, if I add an overload for the **exact** type `List<T>`
```
public static EnumerableAssertions<T> AssertThat<T>(List<T> collection)
{
    return new EnumerableAssertions<T>(collection);
}
```
it works just fine. But of course I don't want to add specific overloads for every type inheriting `IEnumerable<T>`. What also works just fine is calling it like this:
```
List<any-value-type-or-reference-type> list = null;
AssertThat(list.AsEnumerable());
```
Probably because he now has the **exact** type `IEnumerable<T>` again. Is there any chance I can achieve this using type inference whithout having to add explicit overloads for `List<T>`, `T[]`, etc. or having to cast all those using `AsEnumerable()`?

**Comment by RikkiGibson** [comment]
I think what is happening is:

1. method type argument inference occurs.
2. then overload resolution occurs.

After (1), the signature of `AssertThat<T>(T)` is now `AssertThat<List<Something>>(List<Something)`, and `AssertThat<T>(IEnumerable<T>)` is now `AssertThat<Something>(IEnumerable<Something>)`.

[Overload resolution rules](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#11643-better-function-member) states

> if `Mᵥ` has more specific parameter types than `Mₓ`, then `Mᵥ` is better than `Mₓ`.

I'm not sure how else you are using the `T`, but perhaps changing `AssertThat<T>(T)` to `AssertThat(object)` would help here.

**Reply by oliversaggau** [reply]
That explanation makes sense and also explains why adding exact overloads works :bulb: I tried adding a different method
```
public static EnumerableAssertions<T> AssertThat<T, TValue>(T collection) where T : IEnumerable<TValue>
{
    return new EnumerableAssertions<T>(collection);
}
```
whose signature after method type argument inference should become `AssertThat<List<Something>, Something>(List<Something>)` but he still favors `AssertThat<T>(T)` or (after inference) `AssertThat<List<Something>>(List<Something)`. Is this because the former (new method) has more type arguments than the latter? Because the parameters now should be equal sepecific, shouldn't they?

**Reply by oliversaggau** [reply]
As for the use of `T`, I use it In my `ObjectAssertions` in an `Expression` to have something similar to [AssertJ extracting methods](http://joel-costigliola.github.io/assertj/core-8/api/org/assertj/core/api/AbstractObjectAssert.html#extracting-java.util.function.Function...-)
```
public BooleanAssertions Extracting(Expression<Func<T, bool>> expression)
{
	return AssertThat(GetValue(expression));
}

[...]

private TResult GetValue<TResult>(Expression<Func<T, TResult>> expression)
{
	MemberExpression memberExpression = expression.Body as MemberExpression;
	if (memberExpression == null) throw new ArgumentException(string.Format("The expression '{0}' is not a MemberExpression.", expression.ToString()));

	switch (memberExpression.Member)
	{
		case FieldInfo fieldInfo: return (TResult)fieldInfo.GetValue(Target);
		case PropertyInfo propertyInfo: return (TResult)propertyInfo.GetValue(Target);
		default: throw new ArgumentException(string.Format("The expression '{0}' does not refer to a field or property.", expression.ToString()));
	}
}
```
Changing to `AssertThat(object)` would eliminate this.

**Reply by svick** [reply]
> Is this because the former (new method) has more type arguments than the latter?

No, it's because [type inference doesn't work for the new method in the first place](https://sharplab.io/#v2:C4LgbgpgTglgZgTwDQBMQGoA+ABADAAmwEYAWAbgFgAoagGRgGdgAeGAO2AD58AbR4fAF58bCAHcAFAEpKNKgEEGDaMAYA6RcqjAAKgAsAhsAl8mM6tWIA2QgCZ8mlQ2oBvavg+EAzISI3sJA5KKvpGzDpI+DoAagY8AK4QnBI6+ADGAPY8PBBpwDAZbFL4YnrQEFH4IL5e4bEJSe6eLvhNHgC+1O1AA).

---

One (fairly ugly) approach to make this work is [to use inheritance](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUKgBgAJVMAWAbjzwBkBLAZ2AB46pgA+IgG0eCIC8RKAFMA7gAoAlFVwB7AEYArEQGN+ipYOHjpsvAEEGDEWGAMAdEZNmAKgAsAhsAm8mMogHpPJTAGZmWw5DY1NzK1C7JxdND28iW2p8dCIAIUcTazCGPABvPCJCkj9fADYSMiIsqOdAjglbIgA3R24IESkiMXtTEQSiEBJ0AqL83CKJ3wBOCQAiW1mZEcKAXzw13DwMKsjzAbSMkWrzPOXisoqdm2AHZwlSYha2jrOxiY2z1BLSctRK49uLjY/Ce7SkrzOH3GRS+Fz+VzCgLq938dSIqjk3G4amAdDkUHB0MKb0mhVIM1mpACQUWsne6yAA), since methods from a derived class take precedence over better matching methods from a base class.
